[2743] Migrate internal use of CircuitBreaker double to duration

mjmoore opened the pull request:
Fixes #2743 

 * Changes internal usage of `Double` to `Duration`.
 * No changes made to public factory methods (`CircuitBreaker::of`). Questionable to accept a `Duration`, convert it to `Double` and then covert it back to a `Duration` again to pass to the internal constructor. Would appreciate some guidance here on how to manage the `@ExperimentalTime` annotation here.

nomisRev commented:
@mjmoore thanks for this awesome contribution, and congrats on your first contribution 👏 👏 👏 🥳

The CI is currently failing because of some changes in the public API, can you run `./gradlew apiDump` and push the changes? There are more details on this [here](https://github.com/arrow-kt/arrow/blob/main/CONTRIBUTING.md#requirements-to-add-a-new-feature) if you're interested.

I think we should also deprecate the `Double` constructor, so users will be able to automatically refactor.
Full example:

```kotlin
@Deprecated(
   "Prefer the kotlin.time.Duration constructor instead",
    ReplaceWith(
      "of(maxFailures, resetTimeoutNanos.nanoseconds, exponentialBackoffFactor, maxResetTimeout, onRejected, onClosed, onHalfOpen, onOpen)",
      "import kotlin.time.Duration.Companion.nanoseconds"
    )
)
```

>  Would appreciate some guidance here on how to manage the @ExperimentalTime annotation here.

The `kotlin.time.Duration` was stabilised in `1.6.0` so we should be able to simply remove the annotations. [1.6.0 release docs](https://kotlinlang.org/docs/whatsnew16.html#stable-duration-api), or is it requiring `@ExperimentalTime` for something else in the `kotlin.time.*` package?

mjmoore commented:
@nomisRev Thanks for the guidance! Should look a little better now :slightly_smiling_face:

nomisRev commented on arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/CircuitBreaker.kt:
```diff
@@ -612,15 +619,16 @@ private constructor(
-        onOpen
+      CircuitBreaker(
+        state = AtomicRef(Closed(0)),
+        maxFailures = requireNotNull(maxFailures.takeIf { it >= 0 }) { "maxFailures expected to be higher than 0" },
```
```suggestion
        maxFailures = requireNotNull(maxFailures.takeIf { it >= 0 }) { "maxFailures expected to be greater than or equal to 0, but was $maxFailures" },
```

nomisRev commented on arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/CircuitBreaker.kt:
```diff
@@ -612,15 +619,16 @@ private constructor(
+      CircuitBreaker(
+        state = AtomicRef(Closed(0)),
+        maxFailures = requireNotNull(maxFailures.takeIf { it >= 0 }) { "maxFailures expected to be higher than 0" },
+        resetTimeout = requireNotNull(resetTimeout.takeIf { it.isPositive() }) { "resetTimeoutNanos expected to be higher than 0" },
```
```suggestion
        resetTimeout = requireNotNull(resetTimeout.takeIf { it.isPositive() && it != Duration.ZERO }) { "resetTimeoutNanos expected to be greater than ${Duration.ZERO}, but was $resetTimeout" },
```

nomisRev commented on arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/CircuitBreaker.kt:
```diff
@@ -612,15 +619,16 @@ private constructor(
+        state = AtomicRef(Closed(0)),
+        maxFailures = requireNotNull(maxFailures.takeIf { it >= 0 }) { "maxFailures expected to be higher than 0" },
+        resetTimeout = requireNotNull(resetTimeout.takeIf { it.isPositive() }) { "resetTimeoutNanos expected to be higher than 0" },
+        exponentialBackoffFactor = requireNotNull(exponentialBackoffFactor.takeIf { it > 0 }) { "exponentialBackoffFactor expected to be higher than 0" },
```
```suggestion
        exponentialBackoffFactor = requireNotNull(exponentialBackoffFactor.takeIf { it > 0 }) { "exponentialBackoffFactor expected to be greater than 0, but was $exponentialBackoffFactor" },
```

nomisRev commented on arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/CircuitBreaker.kt:
```diff
@@ -612,15 +619,16 @@ private constructor(
+        maxFailures = requireNotNull(maxFailures.takeIf { it >= 0 }) { "maxFailures expected to be higher than 0" },
+        resetTimeout = requireNotNull(resetTimeout.takeIf { it.isPositive() }) { "resetTimeoutNanos expected to be higher than 0" },
+        exponentialBackoffFactor = requireNotNull(exponentialBackoffFactor.takeIf { it > 0 }) { "exponentialBackoffFactor expected to be higher than 0" },
+        maxResetTimeout = requireNotNull(maxResetTimeout.takeIf { it.isPositive() }) { "maxResetTimeout expected to be higher than 0" },
```
```suggestion
        maxResetTimeout = requireNotNull(maxResetTimeout.takeIf { it.isPositive() && it != Duration.ZERO }) { "maxResetTimeout expected to be greater than ${Duration.ZERO}, but was $maxResetTimeout" },
```

nomisRev reviewed (commented):
Looks great to me @mjmoore. Nice work 👏 👏 👏

A couple suggestions for improving the error messages for the preconditions of the constructors. Also it was missing `&& it != Duration.ZERO` check. If you agree with my suggestions, I think it's ready to be merged 🙌 

So this is probably also missing tests. Would be great if you don't mind adding one, but not a blocker for this PR ;)

mjmoore commented:
@nomisRev Thanks for the comments!
 * Updated validation checks.
 * Added some basic tests for constructor validation.
 * Removed duplicated validation. Validation is now performed only in the primary `CircuitBreaker::of` constructor.
 * Reworked the terse validation logic to something a bit more readable.

i-walker commented on arrow-libs/fx/arrow-fx-coroutines/api/arrow-fx-coroutines.api:
```diff
@@ -76,18 +76,18 @@ public final class arrow/fx/coroutines/CircuitBreaker$State$Closed : arrow/fx/co
+	public synthetic fun <init> (JLkotlin/jvm/internal/DefaultConstructorMarker;)V
 	public fun equals (Ljava/lang/Object;)Z
-	public final fun getResetTimeoutNanos ()D
+	public final fun getResetTimeout-UwyO8pc ()J
```
can we redump the api file, just to check if that hash goes away?

i-walker reviewed (dismissed):
Thanks @mjmoore 🙌🏾

nomisRev commented on arrow-libs/fx/arrow-fx-coroutines/src/commonTest/kotlin/arrow/fx/coroutines/CircuitBreakerTest.kt:
```diff
@@ -267,9 +270,43 @@ class CircuitBreakerTest : ArrowFxSpec(
+      ConstructorValues(maxResetTimeout = Duration.ZERO),
+      ConstructorValues(maxResetTimeout = (-1).seconds),
+    ).forEach { value ->
+      "should require valid constructor values" {
```
Awesome!

nomisRev reviewed (approved):
Thank you for the contribution @mjmoore! 🙏 Was my pleasure to work together ;)

nomisRev commented:
@mjmoore Sorry for the long wait. I had a busy week, but I was finally able to dedicate some time to this again.

We should really maintain binary compatibility so we don't break libraries depending on older 1.x.x versions of Arrow,
but it was trickier than I thought with these `@JvmInline value class`.
I was able to maintain binary compatibility by shuffling around some constructors, and properties.

Feel free to copy the changes, or cherry-pick this commit whatever makes most sense for you 👍 
https://github.com/arrow-kt/arrow/commit/1c0bfe6c21608505cb4f3ae61d5f63e509d52b50

nomisRev commented:
PS: @i-walker this fixes the binary compat issues. Thanks for watching out for this 🙏

mjmoore commented:
Hey @nomisRev, @i-walker, finally coming back to this. Cherry picked https://github.com/arrow-kt/arrow/commit/1c0bfe6c21608505cb4f3ae61d5f63e509d52b50, should be good to go now hopefully.

nomisRev commented:
Hey @mjmoore, awesome! 🙌 I'll review it asap 😊

nomisRev commented on arrow-libs/fx/arrow-fx-coroutines/api/arrow-fx-coroutines.api:
```diff
@@ -77,16 +77,20 @@ public final class arrow/fx/coroutines/CircuitBreaker$State$Closed : arrow/fx/co
 
 public final class arrow/fx/coroutines/CircuitBreaker$State$HalfOpen : arrow/fx/coroutines/CircuitBreaker$State {
 	public fun <init> (D)V
+	public synthetic fun <init> (JLkotlin/jvm/internal/DefaultConstructorMarker;)V
 	public fun equals (Ljava/lang/Object;)Z
+	public final fun getResetTimeout-UwyO8pc ()J
 	public final fun getResetTimeoutNanos ()D
 	public fun hashCode ()I
 	public fun toString ()Ljava/lang/String;
 }
 
 public final class arrow/fx/coroutines/CircuitBreaker$State$Open : arrow/fx/coroutines/CircuitBreaker$State {
 	public fun <init> (JD)V
+	public synthetic fun <init> (JJLkotlin/jvm/internal/DefaultConstructorMarker;)V
 	public fun equals (Ljava/lang/Object;)Z
 	public final fun getExpiresAt ()J
+	public final fun getResetTimeout-UwyO8pc ()J
```
All fixed @mjmoore 🙌 

cc\\ @i-walker

i-walker reviewed (approved):
Thanks @mjmoore 🙌🏾

mjmoore commented:
Hey @nomisRev, looks like the previous build failed due to a timeout when configuring `arrow-optics`. I've rebased in case there were any issues with `main`. Would be great if you could kick off a new build.

nomisRev commented:
Hey @mjmoore,
Yes, that is related to a Gradle bug that will be fixed in the next version.

https://github.com/gradle/gradle/issues/19446#event-7377915648.

nomisRev reviewed (approved):
Thank you so much for @mjmoore for your first-contribution 🥳
Really nice work venturing in Arrow Fx straight away 🙌 
Apologies that it took so long to get your contribution in, and thanks for your patience 🙏
